ftw/linkchecker/command/checking_links.py

```diff
@@ -19,31 +27,159 @@ def get_plone_sites_information(app):
     return plone_sites_information
 
 
-def setup_plone(plone_site):
+def setup_plone(app, site_info):
+    plone_site = app.get(site_info['id'])
     user = AccessControl.SecurityManagement.SpecialUsers.system
     user = user.__of__(plone_site.acl_users)
     newSecurityManager(plone_site, user)
     setSite(plone_site)
 
 
-def get_relation_catalog_for_plone_site(app, PloneSiteId):
-    plone_site = app.get(PloneSiteId)
-    setup_plone(plone_site)
-
-    return getUtility(ICatalog)
+def get_broken_relations_in_plone_page():
+    relation_catalog = getUtility(ICatalog)
+    catalog_lookup(relation_catalog)
+    # there's need to also check another catalog
 
 
 def catalog_lookup(relation_catalog):
     x = relation_catalog
     relations = list(x.findRelations())
-    print(relations)
+    relations_info = []
+    for relation in relations:
+        if relation.isBroken():
+            rel_info = {
+                'origin': relation.from_path,
+                'title': relation.from_object.Title()
+            }
+        relations_info.append(rel_info)
+    return relations_info
+
+
+def get_external_broken_links_in_plone_page(site_info):
+    portal_catalog = api.portal.get_tool('portal_catalog')
+    query = {'path': site_info['path'], 'portal_type': 'Document'}
+    brains = portal_catalog.unrestrictedSearchResults(query)
+    links_on_a_plone_site = []
+    for brain in brains:
+        links_on_a_plone_site.append(find_links_on_brain_fields(brain))
+
+    in_field_links = [j for sub in links_on_a_plone_site for j in sub]
+    # check these links and validate
+    # this feature is developed in branch mo/external_link_fetcher
+    # presumably called: work_through_urls(in_field_links)
+
+
+def extract_links_in_string(inputString):
+    # Explanation for the following regex pattern can be found in this post:
+    # https://stackoverflow.com/questions/839994/extracting-a-url-in-python/50790119#50790119
+    regex = ur"\b((?:https?://)?(?:(?:www\.)?(?:[\da-z\.-]+)\.(?:[a-z]{2,6})|(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)|(?:(?:[0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|(?:[0-9a-fA-F]{1,4}:){1,7}:|(?:[0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|(?:[0-9a-fA-F]{1,4}:){1,5}(?::[0-9a-fA-F]{1,4}){1,2}|(?:[0-9a-fA-F]{1,4}:){1,4}(?::[0-9a-fA-F]{1,4}){1,3}|(?:[0-9a-fA-F]{1,4}:){1,3}(?::[0-9a-fA-F]{1,4}){1,4}|(?:[0-9a-fA-F]{1,4}:){1,2}(?::[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:(?:(?::[0-9a-fA-F]{1,4}){1,6})|:(?:(?::[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(?::[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(?:ffff(?::0{1,4}){0,1}:){0,1}(?:(?:25[0-5]|(?:2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(?:25[0-5]|(?:2[0-4]|1{0,1}[0-9]){0,1}[0-9])|(?:[0-9a-fA-F]{1,4}:){1,4}:(?:(?:25[0-5]|(?:2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(?:25[0-5]|(?:2[0-4]|1{0,1}[0-9]){0,1}[0-9])))(?::[0-9]{1,4}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])?(?:/[\w\.-]*)*/?)\b"
```

Comment:
This regex is way better than many I could find or imagine. But we still might look for better options, since nobody can maintain this one.
Nick suggested to lock at [these](https://mathiasbynens.be/demo/url-regex). Many of them aren't much easier to understand though.